proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+   * Pros: No runtime overhead; can trace through foreign code
+   * Cons: Highly platform-specific (currently only available on Linux); slow
+     backtrace collection; imprecise backtraces; large binary size overhead
+     (built with ``-g3``)
+* Info-table provenance (IPE) information (via ``GHC.Stack.CloneStack``):
+   * Pros: Can be used on all platforms; no runtime overhead
+   * Cons: Large binary size overhead; no visibility into foreign code; must be
+     built with ``-finfo-table-map``
+
+All of these backtrace mechanisms have their uses, offering a range of levels
+of detail, executable size, and runtime overhead. Given the complementary
+nature of these mechanisms, GHC should not dictate which of these mechanisms
+should be used to report exception backtraces. Consequently, we use the
+above-described context mechanism to allow backtraces from any of these
+mechanisms to be captured attached to exceptions.
+
+Specifically, we propose to introduce the following set of exception
+annotations: ::
+
+    -- | A backtrace from the cost-centre profiler.
+    data CostCentreBacktrace = CostCentreBacktrace { ... }
+    instance ExceptionAnnotation CostCentreBacktrace
+
+    -- | A backtrace using the native (e.g. DWARF-based) stack unwinder
+    data ExecutionStackBacktrace = ExecutionStackBacktrace { ... }
+    instance ExceptionAnnotation ExecutionStackBacktrace
+
+    -- | A backtrace from GHC's Haskell stack unwinder and info-table
+    -- provenance map.
+    data InfoProvBacktrace = InfoProvBacktrace { ... }
+    instance ExceptionAnnotation InfoProvBacktrace
+
+    -- | A backtrace from HasCallStack evidence.
+    data HasCallStackBacktrace = HasCallStackBacktrace { ... }
+    instance ExceptionAnnotation HasCallStackBacktrace
+
+Handling of rethrowing
+^^^^^^^^^^^^^^^^^^^^^^
+
+One pattern frequently seen in Haskell programs is *rethrowing*. Typically this
+takes the form of catching one type of exception and throwing in its place
+another exception more specific to the application domain. For instance, ::
+
+    data MyAppError = MissingConfigurationError | ...
+
+    readFile "my-app.conf" `catch` $ \ (ioe :: IOError) ->
+        if isDoesNotExistError ioe
+          then throwIO MissingConfigurationError
+          else throwIO ioe
+
+This pattern can be problematic in the presence of exception context: the
+exception thrown by the handler lacks any of the context attached to the
+original ``IOError``, including any backtraces.
+
+While in some select cases dropping context may be desireable (e.g. to avoid
+exposing implementation details unnecessarily to the user), in general this
+proposal seeks to make exception provenance information ubiquitous and
+reliable. Consequently, we propose to that ``catch`` and ``handle`` be modified
+to preserve exception context when an exception is thrown from a handler.
```

Comment:
Can we see what these modifications look like?

In particular I'm wondering if `throwIO` inside `catch` could add an `ExceptionAnnotation` represented as a new "caused by" datatype that stores a `SomeException` (and hence its backtraces/annotations).